### PR TITLE
Topic recognition Pull Request - Siamese s4697173

### PR DESCRIPTION
**Pull request by MatthewCurr**
Hi, This is a pull request to merge the recognition problem for the Siamese Network classification of the ADNI brain dataset. 

The code library is designed for use on the Rangpur cluster by default (Accessing the ADNI dataset from the Rangpur paths) but can be modified to work locally. 

- This library implements the classification of the ADNI dataset into AD and NC labels. 
- A 3D Resnet model was used as the backbone for the Siamese Network. 
- The Siamese Network was implemented with Triplet Loss. 
- The classification model is a simple model comprising several linear layers. 

The following requirements determined by the maintainers have been met:

- Reasonable Commit Log History
- Fork of the topic-recognition branch. 
- Documentation of the problem in the README.md
- Completed test-driver script (predict.py) 

Requires the following libraries: 

- Torch 2.1.0+cu118, 
- pytorch-cuda | 11.7
- seaborn | 0.13.0
- scikit-learn | 1.3.1
- opencv-python | 4.8.1.78
- numpy | 1.26.1
- time | 1.8

Regards,
Matthew (s46971733)

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Good
- Clear flow of information
- Algortihm and usage explained well
- Some plots for loss and 2D manifolds provided (only for the training set for the triplet network though)
- No accuracy plots provided

Commit messages: very good, detailed

Code: 
- uses a triplet network with a 3D resnet backbone
- good design
- well commented

Functionality/Performance: 
- triplet network seem to have separated the two classes very well for the training set, but you could have plotted the loss/accuracy/2D manifolds for the validation set as well to understand issues
- classifier accuracy is only 50% (random guessing?)
- doesn't seem to have performed hyperparameter tuning and no attempts have been made to improve either networks

General comments:
- solves the problem apporpriately although performance is poor
- you need to have observed the performance on the validation set as well for both networks to show/understand potential issues with the algorithm

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem, poor performance <0.6 -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage, accuracy plots would be good -1
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good, no usage -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
